File: 169-majority-element/majority-element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int n = nums.size();
        int majority = 0;

        for (int i = 0; i < 32; ++i) {
            int bitCount = 0;
            for (int num : nums) {
                if ((num >> i) & 1) {
                    bitCount++;
                }
            }

            if (bitCount > n / 2) {
                majority |= (1 << i);
            }
        }

        return majority;
    }
};


// class Solution {
// public:
//     int majorityElement(vector<int>& nums) {
//         int el = -1;
//         int cnt = 0;

//         for(int i = 0; i < nums.size(); i++){
//             if(cnt == 0){
//                 el = nums[i];
//                 cnt++;
//             }
//             else if(nums[i] != el) cnt--;
//             else cnt++;
//         }

//         return el;
//     }
// };

File: 169-majority-element/majority-element.java
class Solution {
    public int majorityElement(int[] nums) {
        int n = nums.length;
        int cnt = 0;
        int el = 0;

        for (int i = 0; i < n; i++) {
            if (cnt == 0) {
                cnt = 1;
                el = nums[i];
            } else if (el == nums[i])
                cnt++;
            else
                cnt--;
        }

        int cnt1 = 0;
        for (int i = 0; i < n; i++) {
            if (nums[i] == el)
                cnt1++;
        }

        if (cnt1 > (n / 2))
            return el;
        return -1;
    }
}